### ISSUE TITLE
Implement Analytics snippet toggle(s) in setup flow if there is an existing tag

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4913]
As part of the existing tag simplification, we are removing the enforcement of any property selection or tag placement. However, it is still crucial that we avoid placing the exact same tag twice on the page. To make this behavior more transparent to the user, and to allow them to still control / override it in these more complicated situations right away, the Analytics setup flow should start showing the toggle(s) to control the snippet placement if there is an existing tag.

In case of no existing tag, the snippet toggles should not be displayed similar to today, since at that point the snippet needs to be placed and there is no point in allowing the user to remove that, it would only break Analytics tracking.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

High-level summary: If there is an existing UA and/or GA4 tag on the site, the Analytics setup flow should render the corresponding toggle(s) to control tag placement, below the respective dropdown, as seen in [this Figma design](https://www.figma.com/file/1VseUbVsR8GYPa83fLkaKx/Existing-tag-UX-improvement?node-id=408%3A208).

### Wording and appearance
**UA `UseSnippetSwitch`**
* Appearance: See [Figma](https://www.figma.com/file/1VseUbVsR8GYPa83fLkaKx/Existing-tag-UX-improvement?node-id=408%3A208), wording is specified below though. The design needs to support both versions, either it's displayed as usual (before white background), or in the blue notice (e.g. needs to use blue text).
* Label: _Place Universal Analytics code_ (like today)
* Note below if selected UA property is the same that the existing UA tag is for (i.e. toggle is off by default): _A tag {existingTagID} for the selected property already exists on the site.  Make sure you remove it if you want to place the same UA tag via Site Kit, otherwise they will be duplicated._ (where `{existingTagID}` is the existing tag `UA-...`, which should match the selected property ID)
* Note below if selected UA property is different from what the existing UA tag is for (i.e. toggle is on by default): _An existing tag {existingTagID} was found on the page. If you prefer to collect data using that existing UA tag, please select the corresponding account and property above._ (where `{existingTagID}` is the existing tag `UA-...`, which should differ from the selected property ID)

**GA4 `UseSnippetSwitch`**
* Appearance: See [Figma](https://www.figma.com/file/1VseUbVsR8GYPa83fLkaKx/Existing-tag-UX-improvement?node-id=408%3A208), wording is specified below though. The design needs to support both versions, either it's displayed as usual (before white background), or in the blue notice (e.g. needs to use blue text).
* Label: _Place Google Analytics 4 code_ (like today)
* Note below if selected GA4 property (or more precisely its web data stream's measurement ID) is the same that the existing GA4 tag is for (i.e. toggle is off by default): _A tag {existingTagID} for the selected property already exists on the site. 
Make sure you remove it if you decide to place the same GA4 tag via Site Kit, otherwise they will be duplicated._ (where `{existingTagID}` is the existing tag `G-...`, which should match the selected property's web data stream measurement ID)
* Note below if selected GA4 property (or more precisely its web data stream's measurement ID) is different from what the existing GA4 tag is for (i.e. toggle is on by default): _An existing tag {existingTagID} was found on the page. If you prefer to collect data using that existing GA4 tag, please select the corresponding account and property above._ (where `{existingTagID}` is the existing tag `G-...`, which should differ from the selected property's web data stream measurement ID)

### Precise conditions
* If there is an existing UA tag and the primary property dropdown is for UA properties (_or_ has a UA property selected, in case of a dropdown for both UA+GA4 properties), the UA `UseSnippetSwitch` should be shown below the three dropdowns.
    * It must be ensured that, if the user selects the same UA property that the existing UA tag is for, the UA `UseSnippetSwitch` is toggled off. If any other UA property is selected, it should be toggled on. The `useExistingTagEffect` should be relied on/enhanced as needed for that.
* If there is an existing GA4 tag and the primary property dropdown is for GA4 properties (_or_ has a GA4 property selected, in case of a dropdown for both UA+GA4 properties), the GA4 `UseSnippetSwitch` should be shown below the two dropdowns.
    * It must be ensured that, if the user selects the same GA4 property that the existing GA4 tag is for (web data stream needs to be checked for that), the GA4 `UseSnippetSwitch` is toggled off. If any other GA4 property is selected, it should be toggled on. A GA4-specific `useExistingTagEffect` should be introduced for that, following similar behavior to the existing UA equivalent.
* If there is an existing UA tag and the blue notice with the UA properties dropdown is shown (i.e. the primary property is a GA4 property and the selected account also has UA properties), the UA `UseSnippetSwitch` should be shown below those two dropdowns in the notice.
    * If the blue notice is just for the "A UA property will be created" information (i.e. no dropdown), the toggle should not be present.
    * It must be ensured that, if the user selects the same UA property that the existing UA tag is for, the UA `UseSnippetSwitch` is toggled off. If any other UA property is selected, it should be toggled on. The `useExistingTagEffect` should be relied on/enhanced as needed for that.
* If there is an existing GA4 tag and the blue notice with the GA4 properties dropdown is shown (i.e. the primary property is a UA property and the selected account also has GA4 properties), the GA4 `UseSnippetSwitch` should be shown below that dropdown in the notice.
    * If the blue notice is just for the "A GA4 property will be created" information (i.e. no dropdown), the toggle should not be present.
    * It must be ensured that, if the user selects the same GA4 property that the existing GA4 tag is for (web data stream needs to be checked for that), the GA4 `UseSnippetSwitch` is toggled off. If any other GA4 property is selected, it should be toggled on. A GA4-specific `useExistingTagEffect` should be introduced for that, following similar behavior to the existing UA equivalent.
* To summarize the above logic: The `UseSnippetSwitch` should only appear if there is an existing tag for the property type (UA or GA4), and it should always appear below the relevant dropdown. In the case that the primary dropdown has both UA and GA4 properties, that means that depending on whether the user selects a UA or GA4 property there, the `UseSnippetSwitch` displayed below would be exchanged to reflect that selection. The most complex possible scenario here is that the selected Analytics account has UA _and_ GA4 properties and there would be an existing UA tag _and_ an existing GA4 tag - in that situation it would mean that both toggles will need to be displayed in the UI.

### Notes on setting up new properties
* If the UA or GA4 property dropdown is displayed and there is an existing tag, the corresponding toggle below should always be shown. That is regardless of what is selected in the property dropdown.
* If the same property that matches the existing tag is selected, the toggle should be off by default (see above).
* If another property (**or "Set up a new property"**) is selected, the toggle should be on by default (see above).
* For now, no special copy should be used for the "Set up a new property" special case, that may be refined separately.

**The PR for this issue should target the [`feature/existing-tag-simplification`](https://github.com/google/site-kit-wp/tree/feature/existing-tag-simplification) branch**

### Create and use GA4 `useExistingTagEffect` hook

- Create the hook in `assets/js/modules/analytics-4/hooks/useExistingTagEffect.js`
- The hook should be based on the existing UA `useExistingTagEffect` in `assets/js/modules/analytics/hooks/useExistingTagEffect.js`, with the code mostly the same.
- Select values from the `MODULES_ANALYTICS_4` store, instead of `MODULES_ANALYTICS`.
- The GA4 measurement ID, retrieved via `getMeasurementID`, should be used in place of `propertyID`.
- Add a call to the new hook at the same point where the UA version is already called in `SetupMain`.
  https://github.com/google/site-kit-wp/blob/c701eeccfb058c225f1f31f74e776875ed9af4ba/assets/js/modules/analytics/components/setup/SetupMain.js#L94

### Create and render `UseSnippetSwitch` component variants

As the text for `UseSnippetSwitch` needs to be changed for the usage in the Setup flow, but remain the same for the current usage in the Settings flow, we are going to create two variants, `SetupUseSnippetSwitch` and `SettingsUseSnippetSwitch` for use in each case.

In each of the following scenarios, to check for an existing tag:

- If the `getExistingTag` selector is being used, check the return value according to the logic seen in `hasExistingTag` to check if there is an existing tag.
- Otherwise, use the `hasExistingTag` selector to check if there is an existing tag.
- Select from the `MODULES_ANALYTICS` store for a UA tag, or the `MODULES_ANALYTICS_4` store for a GA4 tag.

#### Create the new components

For the UA `UseSnippetSwitch` component:

- Refactor `UseSnippetSwitch`, removing the logic to display the text, with the text passed in as a prop instead.
- Create `SettingsUseSnippetSwitch` in the same folder, with the text logic extracted from `UseSnippetSwitch`. Render `UseSnippetSwitch` and pass the text in as a prop.
- Replace the existing use of `UseSnippetSwitch` with `SettingsUseSnippetSwitch`.
- Create `SetupUseSnippetSwitch` in the same folder. Refer to the AC for the text to be displayed depending on whether the selected property ID matches the existing tag property ID. Render `UseSnippetSwitch` and pass the text in as a prop.
  - Use `getPropertyID` and `getExistingTag` to select the appropriate values from the `MODULES_ANALYTICS` store.

Repeat the above for the GA4 `UseSnippetSwitch`, substituting `MODULES_ANALYTICS_4` for `MODULES_ANALYTICS` and measurement ID for property ID.

#### Render Switches in Setup flow

The UA and/or GA4 `SetupUseSnippetSwitch` components need to be rendered in four different components: `SetupFormLegacy`, `SetupFormUA`, `SetupFormGA4`, `SetupFormGA4Transitional`:
https://github.com/google/site-kit-wp/blob/c701eeccfb058c225f1f31f74e776875ed9af4ba/assets/js/modules/analytics/components/setup/SetupForm.js#L97-L102

##### `SetupFormLegacy`

- If there is an existing UA tag, render the UA `SetupUseSnippetSwitch` component after the existing controls.

##### `SetupFormUA`

- If there is an existing UA tag, render the UA `SetupUseSnippetSwitch` component after the existing controls, before the conditional block that renders the `GA4PropertyNotice` component.

##### `SetupFormGA4`

- If there is an existing GA4 tag, render the GA4 `SetupUseSnippetSwitch` component after the existing controls, before the conditional block that renders the `GA4PropertyNotice` component.

##### `SetupFormGA4Transitional`

- If the selected property type is UA (`propertyType === PROPERTY_TYPE_UA`) and there is an existing UA tag, render the UA `SetupUseSnippetSwitch` component above the conditional block that renders the `GA4PropertyNotice` component.
- Otherwise, if the selected property type is GA4 (`propertyType === PROPERTY_TYPE_GA4`) and there is an existing GA4 tag, render the GA4 `SetupUseSnippetSwitch` component at the same point.
- Within the `GA4PropertyNotice` block:
  - If the selected property type is GA4 and there is an existing UA tag, render the UA `SetupUseSnippetSwitch` component at the end of the fragment that's conditionally rendered based on `propertyType === PROPERTY_TYPE_GA4`.
  - Otherwise, if the selected property type is UA and there is an existing GA4 tag, render the GA4 `SetupUseSnippetSwitch` component at the end of the fragment that's conditionally rendered based on `propertyType === PROPERTY_TYPE_UA`.
- Remove the `useEffect` hook that sets the `useSnippet` values, as these values will now be managed by the `SetupUseSnippetSwitch` components in conjunction with `useExistingTagEffect`.

https://github.com/google/site-kit-wp/blob/0e8ad77fbf805bb76f306d4a55845bd06bad22e3/assets/js/modules/analytics/components/setup/SetupFormGA4Transitional.js#L95-L98

### Storybook

- Update the corresponding "with tag" stories for the `SetupForm*` components to ensure they work properly and show the switches. This may be a bit time consuming, as some of the stories appear to no longer be working correctly.

![image](https://user-images.githubusercontent.com/18395600/157235621-85c6db33-a85c-45c3-a6f8-50dc183aa9e4.png)

### Test Coverage

- Fix any broken tests. No new tests needed.

## QA Brief

Please QA this issue on the `feature/existing-tag-simplification` branch; this is not yet merged into `develop` or `main`. 

* Manually add UA and GA4 tag snippets to the site's `header.php`.
* Start the Analytics Setup flow.

Note: Ensure the _stream URL_ matches the site's domain for GA4 tags, otherwise Site Kit won't recognise the tag's measurement ID.
![image](https://user-images.githubusercontent.com/18395600/158663743-f00e2a7b-d0dd-430b-b62a-76996369d0cf.png)

### UA-only flow

* Select an account with only UA properties.
* Verify that the snippet switch appears below the dropdowns as per the AC.
* Ensure the snippet switch toggles **off** when selecting a property which **does** match the existing tag.
* Ensure the snippet switch toggles **on** when selecting a property which **does not** match the existing tag.
* Ensure the snippet switch can still be toggled manually.
* Verify the snippet text conforms to the AC.

Example:
![image](https://user-images.githubusercontent.com/18395600/158660256-b1cace79-b3d1-400c-80a9-36b77565cc75.png)

### GA4-only flow

* Select an account with only GA4 properties.
* Verify that the snippet switch appears below the dropdowns as per the AC.
* Ensure the snippet switch toggles **off** when selecting a property which **does** match the existing tag.
* Ensure the snippet switch toggles **on** when selecting a property which **does not** match the existing tag.
* Ensure the snippet switch can still be toggled manually.
* Verify the snippet text conforms to the AC.

Example:
![image](https://user-images.githubusercontent.com/18395600/158660328-e95904b6-5ffc-4742-8cf0-04056524b41e.png)

### Combined UA and GA4 flow

* Select an account with both UA and GA4 properties.
* Verify that the snippet switches appear below the dropdowns as per the AC.
* Ensure the snippet switches toggle **off** when selecting a property which **does** match the existing tag.
* Ensure the snippet switches toggle **on** when selecting a property which **does not** match the existing tag.
* Ensure the snippet switches can still be toggled manually.
* Verify the snippet text conforms to the AC.

Example:
![image](https://user-images.githubusercontent.com/18395600/158660399-e031dc58-67b5-4032-9bac-61c1c97fa58a.png)

### Storybook

* Check the "with tag" stories for the `SetupForm*` components to ensure they work properly and show the switches.

![image](https://user-images.githubusercontent.com/18395600/157235621-85c6db33-a85c-45c3-a6f8-50dc183aa9e4.png)

## Changelog entry

* Include snippet toggle in Analytics setup flow whenever there is an existing tag.

**Comment by felixarntz** [comment]
@kuasha420 @techanvil @tofumatt Since you've been quite involved with the existing tag simplification work, I wanted to ping you here so you can already review the ACs, in case there are any open questions.

I will still need to define the wording with @marrrmarrr, but the ACs are otherwise already good to go, so let's get a headstart on that. Feel free to already start an IB based on them.

**Comment by techanvil** [comment]
Thanks @felixarntz, I have reviewed the ACs, all seems to be explained, so I don't have any questions at this moment and will proceed to work on the IB. Cheers

**Comment by techanvil** [comment]
Hi @felixarntz, I do now have a question on this. Given that the tag inclusion will now be user-driven via the switches, and also the switch components provides the visual clue plus text to explain the tag will or won't be included, e.g.:

![image](https://user-images.githubusercontent.com/18395600/157101858-16c5b28a-4b70-4055-b22e-bbd993786842.png)

It seems that we should also consider the messaging in [`ExistingTagNotice`](https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics/components/common/ExistingTagNotice.js). The variants where the notice matches the regex `Site Kit will not place its own tags? and rely on the existing ones?`, no longer seem applicable, as the user can override the choice of placing tags. However, the `ExistingTagNotice` component is used in a number of places so it might be more complicated than simply removing these variants. Have these notices already been considered/covered elsewhere?

**Comment by techanvil** [comment]
Hi @felixarntz, I do now have a question on this issue. Reading the AC suggests that when there is an existing tag and the user chooses to create a new property from the dropdown, we should show the `UseSnippetSwitch` for that tag type.

In other words, in this scenario, there should be two switches on the page, for UA and GA4 respectively:

![image](https://user-images.githubusercontent.com/18395600/157231717-c5128cb5-c79d-4ba5-bf2c-87198608f0df.png)

I just want to check this is as intended. As the user has chosen to create a new property, does it make sense to also show the toggle to not actually use it? I don't hold a strong opinion on this, it's adding extra functionality that could be useful, at the same time it might be a bit confusing.

The AC points of _not_ showing the toggle when a property will be created also gave me pause for thought here. It's just, the AC specifically mentions this applies when there is no dropdown, whereas the scenarios I have mentioned above do have dropdowns present.

> If the blue notice is just for the "A UA property will be created" information (i.e. no dropdown), the toggle should not be present.

Anyway, figured I'd better raise this, if you could confirm or clarify the above, that would be appreciated. Cheers!

**Comment by techanvil** [comment]
Further on the above, in the meantime I have fleshed out the IB as much as I can, and added an initial estimate.

**Comment by felixarntz** [comment]
@techanvil You're raising a great point, we had not considered the "Set up a new property" scenario yet. In principle, I agree that the toggles should not be displayed if the user chooses that. However, we are also planning to use the text under the toggles as the main indicators for any existing tag being present in the future, so in order to not increase the scope of this issue, let's do it like this:
* If the user chooses to "Set up a new property", it should be handled exactly the same as if they chose a property that was different from the existing tag.
* In other words, the toggle should remain visible.
* The toggle should be toggled on by default (this _may_ require a tweak to the `useExistingTagEffect`s, not entirely sure from the top of my head though).
* The text should also match what would be shown if they chose a property different from the existing tag.

I've added a brief point to the end of the ACs to clarify that. I will now go ahead and open a separate issue to think about this further and potentially refine, which I'll look into next week with @marrrmarrr.

**Comment by felixarntz** [comment]
@techanvil A few notes on the IB:

> In each of the following scenarios:
>
> - Use the `hasExistingTag` selector from the `MODULES_ANALYTICS` store to check if there is an existing UA tag.
> - Use the `hasExistingTag` selector from the `MODULES_ANALYTICS_4` store to check if there is an existing GA4 tag.

I believe these selectors shouldn't be needed. `!! getExistingTag` would basically be the same, so we can use a single selector for that. (More precisely, we need to of course consider a potential `undefined` value as always. The precise value that would tell us that there is _no_ existing tag is `getExistingTag() === null`.)

> #### Update text in `UseSnippetSwitch` components
>
> * In the UA `UseSnippetSwitch` component, update the conditions so the current text that's displayed when `useSnippet` is `true` is only displayed when there is not an existing UA tag.
> * Otherwise, if there _is_ an existing UA tag and `useSnippet` is `true`, display the text according to the AC depending on whether the selected property ID matches the existing tag property ID.
>   * Use `getPropertyID` and `getExistingTag` to select the appropriate values from the `MODULES_ANALYTICS` store.
> * Repeat the above for the GA4 `UseSnippetSwitch`, substituting `MODULES_ANALYTICS_4` for `MODULES_ANALYTICS` and measurement ID for property ID.

Ahh, I see there may be some confusion here, partly due to my ACs not considering that the `UseSnippetSwitch`s are also used in settings. We shouldn't tamper with the text that is currently present in the components, which is for usage in the Analytics settings. Potentially a better solution here would be to change `UseSnippetSwitch` so that the text is fully controlled via props and then maybe introduce `SetupUseSnippetSwitch` and `SettingsUseSnippetSwitch` variants of it that are prop-less and pass in the correct texts to the `UseSnippetSwitch` (the first based on the ACs here, the latter based on the existing behavior). That would be one possible idea, I'll leave that up to you.

To summarize: The toggle used in Analytics setup should always use wording from the ACs here, while the toggle used in Analytics settings should never use wording from the ACs here but retain its current wording.

> ##### `SetupFormGA4Transitional`
>
> - If the selected property type is UA (`propertyType === PROPERTY_TYPE_UA`) and there is an existing UA tag, render the UA `UseSnippetSwitch` component above the conditional block that renders the `GA4PropertyNotice` component.
> - Otherwise, if the selected property type is UA (`propertyType === PROPERTY_TYPE_UA`) and there is an existing GA4 tag, render the GA4 `UseSnippetSwitch` component at the same point.
> - Within the `GA4PropertyNotice` block:
>   - If the selected property type is GA4 and there is an existing UA tag, render the UA `UseSnippetSwitch` component at the end of the fragment that's conditionally rendered based on `propertyType === PROPERTY_TYPE_GA4`.
>   - Otherwise, if the selected property type is UA and there is an existing GA4 tag, render the GA4 `UseSnippetSwitch` component at the end of the fragment that's conditionally rendered based on `propertyType === PROPERTY_TYPE_UA`.

Maybe I'm misunderstanding the IB here, but this read slightly off to me regarding the placement of the toggles:
* The UA toggle should _always_ be placed right below the dropdown for the UA property. So if the primary dropdown has a UA property selected, the UA toggle should appear there like you say in the first bullet point here.
* However, if there is no existing UA tag but an existing GA4 tag, the GA4 toggle shouldn't appear below the primary dropdown (as it has a UA property selected in that scenario). Because there is no existing UA tag, there should be no toggle below the primary dropdown (i.e. the three dropdowns), and the GA4 toggle should be displayed in the blue box below the dropdown which is specifically for GA4 properties. So your second bullet point here reads off to me in that regard.
* The same applies also for the reverse case, but summarizing: If the UA toggle is relevant, it should always appear right below the UA property dropdown (or the combined dropdown where the UA property is selected). Same for GA4, if the GA4 toggle is relevant, it should always appear right below the GA4 property dropdown (or the combined dropdown where the GA4 property is selected).

> - Remove the `useEffect` hook that sets the `useSnippet` values, as these values will now be managed by the `UseSnippetSwitch` components.
https://github.com/google/site-kit-wp/blob/0e8ad77fbf805bb76f306d4a55845bd06bad22e3/assets/js/modules/analytics/components/setup/SetupFormGA4Transitional.js#L95-L98

Do you mean the `useExistingTagEffect`s here? Setting the value of the `useSnippet` settings shouldn't happen in the `UseSnippetSwitch` components, but in these hooks.

Let me know if these points make sense or if you have any follow-up questions.

**Comment by techanvil** [comment]
Hi @felixarntz, thanks for the notes, you've made some great points here.

> > In each of the following scenarios:
> > 
> > * Use the `hasExistingTag` selector from the `MODULES_ANALYTICS` store to check if there is an existing UA tag.
> > * Use the `hasExistingTag` selector from the `MODULES_ANALYTICS_4` store to check if there is an existing GA4 tag.
> 
> I believe these selectors shouldn't be needed. `!! getExistingTag` would basically be the same, so we can use a single selector for that. (More precisely, we need to of course consider a potential `undefined` value as always. The precise value that would tell us that there is _no_ existing tag is `getExistingTag() === null`.)

This is a valid point - however, I do see some value in `hasExistingTag`, as it actually codifies and thus documents the logic that you've mentioned, rather than having to remember it. I should also mention that I was referring to the example seen in `ExistingTagNotice`, where both `hasExistingTag` and `getExistingTag` are used together (see below), I therefore assumed the more explicit version would be a preference.

Seeing as `hasExistingTag` exists, unless the idea is to actually deprecate/remove it, I do think it's still an appropriate choice if `getExistingTag` does not need to be used within the same scope.

https://github.com/google/site-kit-wp/blob/887c75f420e361524cdfbc4293279d54efea9205/assets/js/modules/analytics/components/common/ExistingTagNotice.js#L37-L39

With that in mind, my suggestion is to amend the IB text you've highlighted to read as follows:

> In each of the following scenarios:
> 
> * If the `getExistingTag` selector is being used, check the return value according to the logic seen in `hasExistingTag` to check if there is an existing tag.
> * Otherwise, use the `hasExistingTag` selector to check if there is an existing tag.
> * Select from the `MODULES_ANALYTICS` store for a UA tag, or the `MODULES_ANALYTICS_4` store for a GA4 tag.

However, if you'd prefer to remove the use of `hasExistingTag` altogether, please let me know. In the meantime, I've made this change to the IB.

---

> > #### Update text in `UseSnippetSwitch` components
> > 
> > * In the UA `UseSnippetSwitch` component, update the conditions so the current text that's displayed when `useSnippet` is `true` is only displayed when there is not an existing UA tag.
> > * Otherwise, if there _is_ an existing UA tag and `useSnippet` is `true`, display the text according to the AC depending on whether the selected property ID matches the existing tag property ID.
> >   
> >   * Use `getPropertyID` and `getExistingTag` to select the appropriate values from the `MODULES_ANALYTICS` store.
> > * Repeat the above for the GA4 `UseSnippetSwitch`, substituting `MODULES_ANALYTICS_4` for `MODULES_ANALYTICS` and measurement ID for property ID.
> 
> Ahh, I see there may be some confusion here, partly due to my ACs not considering that the `UseSnippetSwitch`s are also used in settings. We shouldn't tamper with the text that is currently present in the components, which is for usage in the Analytics settings. Potentially a better solution here would be to change `UseSnippetSwitch` so that the text is fully controlled via props and then maybe introduce `SetupUseSnippetSwitch` and `SettingsUseSnippetSwitch` variants of it that are prop-less and pass in the correct texts to the `UseSnippetSwitch` (the first based on the ACs here, the latter based on the existing behavior). That would be one possible idea, I'll leave that up to you.
> 
> To summarize: The toggle used in Analytics setup should always use wording from the ACs here, while the toggle used in Analytics settings should never use wording from the ACs here but retain its current wording.

Thanks for pointing this out. I think your suggestion of passing text into `UseSnippetSwitch` as a prop, and creating `SetupUseSnippetSwitch` and `SettingsUseSnippetSwitch` variants makes sense, and have updated the IB accordingly.

> > ##### `SetupFormGA4Transitional`
> > 
> > * If the selected property type is UA (`propertyType === PROPERTY_TYPE_UA`) and there is an existing UA tag, render the UA `UseSnippetSwitch` component above the conditional block that renders the `GA4PropertyNotice` component.
> > * Otherwise, if the selected property type is UA (`propertyType === PROPERTY_TYPE_UA`) and there is an existing GA4 tag, render the GA4 `UseSnippetSwitch` component at the same point.
> > * Within the `GA4PropertyNotice` block:
> >   
> >   * If the selected property type is GA4 and there is an existing UA tag, render the UA `UseSnippetSwitch` component at the end of the fragment that's conditionally rendered based on `propertyType === PROPERTY_TYPE_GA4`.
> >   * Otherwise, if the selected property type is UA and there is an existing GA4 tag, render the GA4 `UseSnippetSwitch` component at the end of the fragment that's conditionally rendered based on `propertyType === PROPERTY_TYPE_UA`.
> 
> Maybe I'm misunderstanding the IB here, but this read slightly off to me regarding the placement of the toggles:
> 
> * The UA toggle should _always_ be placed right below the dropdown for the UA property. So if the primary dropdown has a UA property selected, the UA toggle should appear there like you say in the first bullet point here.
> * However, if there is no existing UA tag but an existing GA4 tag, the GA4 toggle shouldn't appear below the primary dropdown (as it has a UA property selected in that scenario). Because there is no existing UA tag, there should be no toggle below the primary dropdown (i.e. the three dropdowns), and the GA4 toggle should be displayed in the blue box below the dropdown which is specifically for GA4 properties. So your second bullet point here reads off to me in that regard.
> * The same applies also for the reverse case, but summarizing: If the UA toggle is relevant, it should always appear right below the UA property dropdown (or the combined dropdown where the UA property is selected). Same for GA4, if the GA4 toggle is relevant, it should always appear right below the GA4 property dropdown (or the combined dropdown where the GA4 property is selected).

Thanks for raising this - and, apologies, as I've introduced some confusion here due to a copy/paste error. The second point was intended to read as follows:

> - Otherwise, if the selected property type is GA4 (`propertyType === PROPERTY_TYPE_GA4`) and there is an existing GA4 tag, render the GA4 `SetupUseSnippetSwitch` component at the same point.

However, I do believe it's the correct approach to render the GA4 switch below the primary dropdown in that case, as the primary dropdown will have a GA4 property selected, and the blue area will contain the UA dropdown (please see the [relevant code](https://github.com/google/site-kit-wp/blob/887c75f420e361524cdfbc4293279d54efea9205/assets/js/modules/analytics/components/setup/SetupFormGA4Transitional.js#L132-L169) in `SetupFormGA4Transitional` for details).

Therefore, I think that with the above amendment to the second point, the section should be valid. I have addressed this in the IB.

> > * Remove the `useEffect` hook that sets the `useSnippet` values, as these values will now be managed by the `UseSnippetSwitch` components.
> >   https://github.com/google/site-kit-wp/blob/0e8ad77fbf805bb76f306d4a55845bd06bad22e3/assets/js/modules/analytics/components/setup/SetupFormGA4Transitional.js#L95-L98
> 
> Do you mean the `useExistingTagEffect`s here? Setting the value of the `useSnippet` settings shouldn't happen in the `UseSnippetSwitch` components, but in the `useExistingTagEffect` hooks.

I've taken another look at this, but, I do believe this is correct. The `useSnippet` settings currently _are_ managed in the `UseSnippetSwitch` components, and it's still appropriate for them to be, as far as I can see.

https://github.com/google/site-kit-wp/blob/887c75f420e361524cdfbc4293279d54efea9205/assets/js/modules/analytics/components/common/UseSnippetSwitch.js#L48-L50

The calls to `setUseSnippet` in `useExistingTagEffect` are useful as they are, as they result in the snippet being enabled/disabled when the user selects a new property from the dropdown (and thus, will change the position of the related switch, which is determined by the `useSnippet` value).

However, the call to `setUseSnippet` in the `onChange` handler in  `UseSnippetSwitch` is still necessary in order for the user to be able to override the snippet use initially determined when choosing the property.

At present, as far as I can tell, the `useEffect` in `SetupFormGA4Transitional` appears to be redundant anyway, as the `useSnippet` value will be `true`, unless there is an existing tag, in which case it will be managed by `useExistingTagEffect` (in conjunction with `UseSnippetSwitch` once this is implemented). It's worth noting that the other related `SetupForm*` components _don't_ have a similar `useEffect`.

**Comment by felixarntz** [comment]
@techanvil 
> Seeing as `hasExistingTag` exists, unless the idea is to actually deprecate/remove it, I do think it's still an appropriate choice if `getExistingTag` does not need to be used within the same scope.

Definitely, if you don't need to use `getExistingTag` in a component, using `hasExistingTag` makes sense for just a bool check. But if you're already using `getExistingTag`, it's technically unnecessary to also use `hasExistingTag`. Either way it's not a big deal.

> However, I do believe it's the correct approach to render the GA4 switch below the primary dropdown in that case, as the primary dropdown will have a GA4 property selected

Correct, my feedback was based on your copy/paste error version. Now that you've clarified, that sounds right.

> At present, as far as I can tell, the `useEffect` in `SetupFormGA4Transitional` appears to be redundant anyway, as the `useSnippet` value will be `true`, unless there is an existing tag, in which case it will be managed by `useExistingTagEffect` (in conjunction with `UseSnippetSwitch` once this is implemented). It's worth noting that the other related `SetupForm*` components _don't_ have a similar `useEffect`.

Correct. All your reasoning here makes sense, maybe my original comment on this wasn't phrased correctly. My point was that that `useEffect` is more covered by what `useExistingTagEffect` than by what `UseSnippetSwitch` does. So yes, it is redundant based on `useExistingTagEffect` and should be removed.

IB ✅

**Comment by tofumatt** [comment]
Note: I'm not attaching this issue to a release since the change is currently in the feature branch and won't be shipped into `develop`/`main` yet.

**Comment by techanvil** [comment]
Hi @wpdarren, thanks for taking a look at this.

> 1. For Storybook, what will the URL be? Started looking at the screens then realised I was looking at `develop` since this is on `feature/existing-tag-simplification` branch.

Ah, sorry - I forgot this was merged to the `feature/existing-tag-simplification` branch. The Storybook for the related PR can be seen here: https://google.github.io/site-kit-wp/storybook/pull/4893/?path=/story/modules-analytics-setup-setupformga4--with-existing-tag

> 2. This is likely out of scope, but when I go to the settings, there's no message warning users that if they enable the toggle for a property that already exists (i.e. in header.php) that duplicate data will occur. We've added this to the setup flow but not for settings.  Do you think this needs another ticket or is it covered elsewhere?

This is something @felixarntz has clarified on an earlier [comment](https://github.com/google/site-kit-wp/issues/4913#issuecomment-1063257467), we are not updating the text in settings, only for the setup flow. Here's the relevant paragraph from the linked comment:

> Ahh, I see there may be some confusion here, partly due to my ACs not considering that the UseSnippetSwitchs are also used in settings. We shouldn't tamper with the text that is currently present in the components, which is for usage in the Analytics settings. Potentially a better solution here would be to change UseSnippetSwitch so that the text is fully controlled via props and then maybe introduce SetupUseSnippetSwitch and SettingsUseSnippetSwitch variants of it that are prop-less and pass in the correct texts to the UseSnippetSwitch (the first based on the ACs here, the latter based on the existing behavior). That would be one possible idea, I'll leave that up to you.

**Comment by wpdarren** [comment]
Thanks @techanvil 👍 

### QA Update: 

**Verified:**

- Checked the "with tag" stories for the SetupForm components and they work properly and show the switches.

### UA-only flow

- The snippet switches appear below the dropdowns as per the AC.
- The snippet switches toggle off when selecting a property which does match the existing tag.
- The snippet switches toggle on when selecting a property which does not match the existing tag.
- Confirmed that the snippet switches can still be toggled manually.
- The snippet text conforms to the AC.

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/158991902-5ce21c08-c0cd-4409-9326-8ca5fbd5ad9a.png)
![image](https://user-images.githubusercontent.com/73545194/158992349-95c93a13-2895-41f5-9dc6-f4f9b7cae850.png)

</details>

### GA4-only flow

- The snippet switches appear below the dropdowns as per the AC.
- The snippet switches toggle off when selecting a property which does match the existing tag.
- The snippet switches toggle on when selecting a property which does not match the existing tag.
- Confirmed that the snippet switches can still be toggled manually.
- The snippet text conforms to the AC.

<details>
 <summary>Screenshots</summary>
 
![image](https://user-images.githubusercontent.com/73545194/158999426-f67821f1-9df6-46b2-938b-efe4f4ef7d49.png)
![image](https://user-images.githubusercontent.com/73545194/158999821-32f56ef5-b74d-47b8-92da-1c41dff92823.png)

</details>

### Combined UA and GA4 flow

- Snippet switches appear below the dropdowns as per the AC.
- The snippet switches toggle off when selecting a property which does match the existing tag.
- The snippet switches toggle on when selecting a property which does not match the existing tag.
- Confirmed that the snippet switches can still be toggled manually.
- The snippet text conforms to the AC.

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/158986185-0968125c-4412-4c04-bbb8-4b0a61bac7e5.png)
![image](https://user-images.githubusercontent.com/73545194/158986209-eaf1f192-4373-4dff-a4a1-964af6b2bc24.png)
![image](https://user-images.githubusercontent.com/73545194/158987113-2030b040-86d6-4aa3-857f-f8a1ab0d6ee6.png)
![image](https://user-images.githubusercontent.com/73545194/158987148-51e2a09a-b107-4473-bd84-0405a9f6116c.png)

</details>
